--- Day 5: Hydrothermal Venture ---
You come across a field of hydrothermal vents on the ocean floor! These vents constantly produce large, opaque clouds, so it would be best to avoid them if possible.

They tend to form in lines; the submarine helpfully produces a list of nearby lines of vents (your puzzle input) for you to review. For example:

    0,9 -> 5,9
    8,0 -> 0,8
    9,4 -> 3,4
    2,2 -> 2,1
    7,0 -> 7,4
    6,4 -> 2,0
    0,9 -> 2,9
    3,4 -> 1,4
    0,0 -> 8,8
    5,5 -> 8,2

Each line of vents is given as a line segment in the format x1,y1 -> x2,y2 where x1,y1 are the coordinates of one end the line segment and x2,y2 are the coordinates of the other end. These line segments include the points at both ends. In other words:

An entry like 1,1 -> 1,3 covers points 1,1, 1,2, and 1,3.
An entry like 9,7 -> 7,7 covers points 9,7, 8,7, and 7,7.

For now, only consider horizontal and vertical lines: lines where either x1 = x2 or y1 = y2.

So, the horizontal and vertical lines from the above list would produce the following diagram:

    .......1..
    ..1....1..
    ..1....1..
    .......1..
    .112111211
    ..........
    ..........
    ..........
    ..........
    222111....


In this diagram, the top left corner is 0,0 and the bottom right corner is 9,9. Each position is shown as the number of lines which cover that point or . if no line covers that point. The top-left pair of 1s, for example, comes from 2,2 -> 2,1; the very bottom row is formed by the overlapping lines 0,9 -> 5,9 and 0,9 -> 2,9.

To avoid the most dangerous areas, you need to determine the number of points where at least two lines overlap. In the above example, this is anywhere in the diagram with a 2 or larger - a total of 5 points.

Consider only horizontal and vertical lines. At how many points do at least two lines overlap?

In [4]:
!python --version

Python 3.9.7


In [17]:
import numpy as np
import pandas as pd
np.__version__, pd.__version__

('1.20.2', '1.2.3')

In [6]:
with open("day5.txt") as f:
    lines = f.readlines()
lines = [(pts[0].strip(),pts[1].strip()) for l in lines if (pts := l.split("->"))]
lines[:5]

[('561,579', '965,175'),
 ('735,73', '316,73'),
 ('981,566', '981,11'),
 ('631,588', '631,910'),
 ('919,964', '70,115')]

In [7]:
lines = [((t[0].split(",")[0],t[0].split(",")[1]), (t[1].split(",")[0],t[1].split(",")[1])) for t in lines]
lines[0]

(('561', '579'), ('965', '175'))

In [11]:
lines = [(tuple(int(n) for n in line[0]), tuple(int(n) for n in line[1])) for line in lines]
lines[:5]

[((561, 579), (965, 175)),
 ((735, 73), (316, 73)),
 ((981, 566), (981, 11)),
 ((631, 588), (631, 910)),
 ((919, 964), (70, 115))]

In [12]:
len(lines)

500

In [18]:
# https://stackoverflow.com/questions/2158395/flatten-an-irregular-list-of-lists
from collections.abc import Iterable

def flatten(l):
    for el in l:
        if isinstance(el, Iterable) and not isinstance(el, (str, bytes)):
            yield from flatten(el)
        else:
            yield el

In [15]:
flat = list(flatten(lines))
flat[:10]

[561, 579, 965, 175, 735, 73, 316, 73, 981, 566]

In [16]:
min(flat),max(flat)

(10, 989)

In [31]:
height = 1_000
width = 1_000
df = pd.DataFrame(np.zeros((height,width)))
df.shape

(1000, 1000)

In [23]:
df.head()

0    1    2    3    4    5    6    7    8    9    ...  990  991  992  993  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

   994  995  996  997  998  999  
0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 1000 columns]

In [32]:
df = df.astype("int")
df.head()

0    1    2    3    4    5    6    7    8    9    ...  990  991  992  993  \
0    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   

   994  995  996  997  998  999  
0    0    0    0    0    0    0  
1    0    0    0    0    0    0  
2    0    0    0    0    0    0  
3    0    0    0    0    0    0  
4    0    0    0    0    0    0  

[5 rows x 1000 columns]

In [27]:
df.iloc[2:6,4] += 1
df.head(10)

0    1    2    3    4    5    6    7    8    9    ...  990  991  992  993  \
0    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
2    0    0    0    0    1    0    0    0    0    0  ...    0    0    0    0   
3    0    0    0    0    1    0    0    0    0    0  ...    0    0    0    0   
4    0    0    0    0    1    0    0    0    0    0  ...    0    0    0    0   
5    0    0    0    0    1    0    0    0    0    0  ...    0    0    0    0   
6    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
7    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
8    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
9    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   

   994  995  996  997  998  999  
0    0    0    0    0    0    0  
1    0    0    0    0    0    0  
2    0    0    0    0    0    0  
3    0    0    0    0    0    0  
4    0    0    0    0    0    0  
5    0    0    0    0    0    0  
6    0    0    0    0    0    0  
7    0    0    0    0    0    0  
8    0    0    0    0    0    0  
9    0    0    0    0    0    0  

[10 rows x 1000 columns]

In [28]:
for line in lines[:5]:
    (x1,y1),(x2,y2) = line
    print(x1,y1,x2,y2)

561 579 965 175
735 73 316 73
981 566 981 11
631 588 631 910
919 964 70 115


In [30]:
lines[3:4]

[((631, 588), (631, 910))]

In [33]:
d = df.copy()

In [34]:
d = df.copy()
for line in lines:
    (x1,y1),(x2,y2) = line
    if x1 == x2:
        y1, y2 = min(y1,y2), max(y1,y2)
        d.iloc[y1:y2+1,x1] += 1
        continue
    if y1 == y2:
        x1, x2 = min(x1,x2), max(x1,x2)
        d.iloc[y1, x1:x2+1] +=1
        continue
d.head(10)

0    1    2    3    4    5    6    7    8    9    ...  990  991  992  993  \
0    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
5    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
6    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
7    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
8    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
9    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   

   994  995  996  997  998  999  
0    0    0    0    0    0    0  
1    0    0    0    0    0    0  
2    0    0    0    0    0    0  
3    0    0    0    0    0    0  
4    0    0    0    0    0    0  
5    0    0    0    0    0    0  
6    0    0    0    0    0    0  
7    0    0    0    0    0    0  
8    0    0    0    0    0    0  
9    0    0    0    0    0    0  

[10 rows x 1000 columns]

In [35]:
d.sum().sum()

114149

In [36]:
(d >= 2).sum().sum()

6005

In [ ]:
d = df.copy()
for line in lines:
    (x1,y1),(x2,y2) = line
    if x1 == x2:
        y1, y2 = min(y1,y2), max(y1,y2)
        d.iloc[y1:y2+1,x1] += 1
        continue
    if y1 == y2:
        x1, x2 = min(x1,x2), max(x1,x2)
        d.iloc[y1, x1:x2+1] +=1
        continue

Your puzzle answer was 6005.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
Unfortunately, considering only horizontal and vertical lines doesn't give you the full picture; you need to also consider diagonal lines.

Because of the limits of the hydrothermal vent mapping system, the lines in your list will only ever be horizontal, vertical, or a diagonal line at exactly 45 degrees. In other words:

An entry like 1,1 -> 3,3 covers points 1,1, 2,2, and 3,3.
An entry like 9,7 -> 7,9 covers points 9,7, 8,8, and 7,9.

Considering all lines from the above example would now produce the following diagram:

    1.1....11.
    .111...2..
    ..2.1.111.
    ...1.2.2..
    .112313211
    ...1.2....
    ..1...1...
    .1.....1..
    1.......1.
    222111....

You still need to determine the number of points where at least two lines overlap. In the above example, this is still anywhere in the diagram with a 2 or larger - now a total of 12 points.

Consider all of the lines. At how many points do at least two lines overlap?

In [38]:
d = df.copy()
for line in lines:
    (x1,y1),(x2,y2) = line
    if x1 == x2:
        y1, y2 = min(y1,y2), max(y1,y2)
        d.iloc[y1:y2+1,x1] += 1
        continue
    if y1 == y2:
        x1, x2 = min(x1,x2), max(x1,x2)
        d.iloc[y1, x1:x2+1] +=1
        continue
    if x1 < x2:
        xstep = 1
    else:
        xstep = -1
    if y1 < y2:
        ystep = 1
    else:
        ystep = -1
    x,y = x1,y1
    while True:
        d.iloc[y,x] += 1
        x += xstep
        y += ystep
        if x == x2 and y == y2:
            d.iloc[y,x] += 1
            break
(d >= 2).sum().sum()        

23864